### PR TITLE
Add support for local COPC reading

### PR DESCRIPTION
**Pull request by NEDJIMAbelgacem**
## Description
This PR makes use of lazperf 3.0 to read COPC files.
- Link to COPC format document: https://copc.io/

### CODE REVIEW

File: src/core/pointcloud/qgseptdecoder.cpp [hunk 1]
```diff
@@ -672,4 +672,303 @@ QgsPointCloudBlock *QgsEptDecoder::decompressLaz( const QByteArray &byteArrayDat
...
+          break;
+        }
+      }
+      if ( !foundAttr )
+      {
+        // this can possibly happen -- e.g. if a style built using a different point cloud format references an attribute which isn't available from the laz file
+        requestedAttributeDetails.emplace_back( RequestedAttributeDetails( LazAttribute::MissingOrUnknown, requestedAttribute.type(), requestedAttribute.size() ) );
+      }
+    }
+  }
```
Comment (wonder-sk) [inline src/core/pointcloud/qgseptdecoder.cpp:827]: That's a lot of duplicated code from decompressLaz() - let's move the common preparation code to a function.

File: src/core/pointcloud/qgseptdecoder.cpp [hunk 2]
```diff
@@ -672,4 +672,303 @@ QgsPointCloudBlock *QgsEptDecoder::decompressLaz( const QByteArray &byteArrayDat
...
+              _storeToStream<double>( dataBuffer, outputOffset, requestedAttribute.type, *reinterpret_cast<double * >( &buf[requestedAttribute.offset] ) );
+              break;
+          }
+        }
+        break;
+        case LazAttribute::MissingOrUnknown:
+          // just store 0 for unknown/missing attributes
+          _storeToStream<unsigned short>( dataBuffer, outputOffset, requestedAttribute.type, 0 );
+          break;
+      }
```
Comment (wonder-sk) [inline src/core/pointcloud/qgseptdecoder.cpp:951]: Also a lot of duplication here - most of the internal part of the for-loop could go to a separate function to unpack and store data to stream...

File: src/core/pointcloud/qgseptdecoder.cpp [hunk 3]
```diff
@@ -672,4 +672,303 @@ QgsPointCloudBlock *QgsEptDecoder::decompressLaz( const QByteArray &byteArrayDat
...
+        // if the point is filtered out, rewind the offset so the next point is written over it
+        outputOffset -= requestedPointRecordSize;
+        ++skippedPoints;
+      }
+    }
+  }
+
+  block->setPointCount( pointCount - skippedPoints );
+  return block.release();
+}
```
Comment (wonder-sk) [inline src/core/pointcloud/qgseptdecoder.cpp:972]: It would be probably good to move any lazperf-related code to a static functions in a new class (e.g. QgsLazperfUtils) which would not be a part of public API. And COPC would not need QgsEptDecoder class at all, it would only use functions from the new one.

File: src/core/pointcloud/qgspointcloudindex.h [hunk 4]
```diff
@@ -262,7 +262,7 @@ class CORE_EXPORT QgsPointCloudIndex: public QObject
     QgsDoubleRange nodeZRange( const IndexedPointCloudNode &node ) const;
 
     //! Returns node's error in map units (used to determine in whether the node has enough detail for the current view)
-    float nodeError( const IndexedPointCloudNode &n ) const;
+    virtual float nodeError( const IndexedPointCloudNode &n ) const;
```
Comment (wonder-sk) [inline src/core/pointcloud/qgspointcloudindex.h:265]: why is this changed to virtual?

File: src/core/providers/copc/qgscopcprovider.cpp [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+  return false;
+}
+
+QVariantMap QgsCopcProviderMetadata::decodeUri( const QString &uri ) const
+{
+  const QString path = uri;
+  QVariantMap uriComponents;
+  uriComponents.insert( QStringLiteral( "path" ), path );
+  uriComponents.insert( QStringLiteral( "isCopc" ), uri.endsWith( ".copc.laz" ) );
```
Comment (wonder-sk) [inline src/core/providers/copc/qgscopcprovider.cpp:203]: What is the idea with "isCopc" ?

File: src/gui/providers/qgspointcloudsourceselect.cpp [hunk 6]
```diff
@@ -78,6 +78,10 @@ void QgsPointCloudSourceSelect::addButtonClicked()
       // auto determine preferred provider for each path
 
       const QList< QgsProviderRegistry::ProviderCandidateDetails > preferredProviders = QgsProviderRegistry::instance()->preferredProvidersForUri( mPath );
+      for ( QgsProviderRegistry::ProviderCandidateDetails p : preferredProviders )
+      {
+        qDebug() << p.metadata()->key();
+      }
```
Comment (wonder-sk) [inline src/gui/providers/qgspointcloudsourceselect.cpp:84]: remove debug info?

File: src/core/pointcloud/qgscopcpointcloudindex.cpp [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+  attributes.push_back( QgsPointCloudAttribute( "Classification", ( QgsPointCloudAttribute::DataType ) 0 ) );
+  attributes.push_back( QgsPointCloudAttribute( "Intensity", ( QgsPointCloudAttribute::DataType ) 3 ) );
+  attributes.push_back( QgsPointCloudAttribute( "ReturnNumber", ( QgsPointCloudAttribute::DataType ) 0 ) );
+  attributes.push_back( QgsPointCloudAttribute( "NumberOfReturns", ( QgsPointCloudAttribute::DataType ) 0 ) );
+  attributes.push_back( QgsPointCloudAttribute( "ScanDirectionFlag", ( QgsPointCloudAttribute::DataType ) 0 ) );
+  attributes.push_back( QgsPointCloudAttribute( "EdgeOfFlightLine", ( QgsPointCloudAttribute::DataType ) 0 ) );
+  attributes.push_back( QgsPointCloudAttribute( "ScanAngleRank", ( QgsPointCloudAttribute::DataType ) 8 ) );
+  attributes.push_back( QgsPointCloudAttribute( "UserData", ( QgsPointCloudAttribute::DataType ) 0 ) );
+  attributes.push_back( QgsPointCloudAttribute( "PointSourceId", ( QgsPointCloudAttribute::DataType ) 3 ) );
+  attributes.push_back( QgsPointCloudAttribute( "GpsTime", ( QgsPointCloudAttribute::DataType ) 9 ) );
```
[the remark this answers is not in the document]
Comment (NEDJIMAbelgacem) [reply src/core/pointcloud/qgscopcpointcloudindex.cpp:118]: I just printed these from EPT and used them here and forgot to  QgsPointCloudAttribute::DataType directly

File: src/core/pointcloud/qgspointcloudindex.h (hunk 4 above)
Comment (NEDJIMAbelgacem) [reply src/core/pointcloud/qgspointcloudindex.h:265]: I was going to use the span from COPC at some point to calculate node error and forgot to remove "virtual"

File: src/app/qgisapp.cpp [hunk 8]
```diff
@@ -7460,7 +7460,7 @@ bool QgisApp::openLayer( const QString &fileName, bool allowInteractive )
 
   // highest priority = delegate to provider registry to handle
   const QList< QgsProviderRegistry::ProviderCandidateDetails > candidateProviders = QgsProviderRegistry::instance()->preferredProvidersForUri( fileName );
-  if ( candidateProviders.size() == 1 && candidateProviders.at( 0 ).layerTypes().size() == 1 )
+  if ( candidateProviders.size() >= 1 && candidateProviders.at( 0 ).layerTypes().size() == 1 )
   {
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp:7469]: This is not right -- that branch only applies when there's a single candidate

Comment (NEDJIMAbelgacem) [reply src/app/qgisapp.cpp:7469]: Actually .copc.laz file could be read by using both pdal (as a normal laz file) or with the COPC provider. 
So preferredProvidersForUri is returning both providers...

Comment (nyalldawson) [reply src/app/qgisapp.cpp:7469]: You could add .copc.laz to the Blocklist of URIs for the pdal provider.

Comment (NEDJIMAbelgacem) [reply src/app/qgisapp.cpp:7469]: You're right, I put a higher priority for the copc provider (101) and now preferredProvidersForUri returns just one provider.

File: src/core/providers/copc/qgscopcprovider.cpp [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+bool QgsCopcProviderMetadata::uriIsBlocklisted( const QString &uri ) const
+{
+  const QVariantMap parts = decodeUri( uri );
+  if ( !parts.contains( QStringLiteral( "path" ) ) )
+    return false;
+
+  const QFileInfo fi( parts.value( QStringLiteral( "path" ) ).toString() );
+
+  // internal details only
+  if ( parts.value( QStringLiteral( "path" ) ).toString().endsWith( ".copc.laz", Qt::CaseSensitivity::CaseInsensitive ) )
```
Comment (nyalldawson) [inline src/core/providers/copc/qgscopcprovider.cpp:192]: This logic is wrong, here you are saying that a ".copc.laz" file should never be exposed to users.

This should be handled in QgsPdalProviderMetadata instead -- in QgsPdalProviderMetadata::priorityForUri/validLayerTypesForUri/querySubLayers you should adapt the existing tests so that .copc.laz files aren't accepted. Right now QgsPdalProviderMetadata just checks for a .laz extension, so it's greedily grabbing the .copc.laz uris.

**Review by wonder-sk — APPROVED** [review]
It works well and the code looks good! :tada:

Let's merge!
